src/client/Client.cc

```diff
@@ -10241,30 +10246,39 @@ void Client::_release_filelocks(Fh *fh)
   Inode *in = fh->inode.get();
   ldout(cct, 10) << __func__ << " " << fh << " ino " << in->ino << dendl;
 
+  list<ceph_filelock> activated_locks;
+
   list<pair<int, ceph_filelock> > to_release;
 
   if (fh->fcntl_locks) {
     auto &lock_state = fh->fcntl_locks;
-    for(multimap<uint64_t, ceph_filelock>::iterator p = lock_state->held_locks.begin();
-	p != lock_state->held_locks.end();
-	++p)
-      to_release.push_back(pair<int, ceph_filelock>(CEPH_LOCK_FCNTL, p->second));
+    for(auto p = lock_state->held_locks.begin(); p != lock_state->held_locks.end(); ) {
+      auto q = p++;
+      if (in->flags & I_ERROR_FILELOCK) {
+	lock_state->remove_lock(q->second, activated_locks);
+      } else {
+	to_release.push_back(pair<int, ceph_filelock>(CEPH_LOCK_FCNTL, q->second));
+      }
+    }
     lock_state.reset();
   }
   if (fh->flock_locks) {
     auto &lock_state = fh->flock_locks;
-    for(multimap<uint64_t, ceph_filelock>::iterator p = lock_state->held_locks.begin();
-	p != lock_state->held_locks.end();
-	++p)
-      to_release.push_back(pair<int, ceph_filelock>(CEPH_LOCK_FLOCK, p->second));
+    for(auto p = lock_state->held_locks.begin(); p != lock_state->held_locks.end(); ) {
+      auto q = p++;
+      if (in->flags & I_ERROR_FILELOCK) {
+	lock_state->remove_lock(q->second, activated_locks);
+      } else {
+	to_release.push_back(pair<int, ceph_filelock>(CEPH_LOCK_FLOCK, q->second));
+      }
+    }
     lock_state.reset();
   }
 
-  if (to_release.empty())
-    return;
+  if ((in->flags & I_ERROR_FILELOCK) && !in->has_any_filelocks())
+    in->flags &= ~I_ERROR_FILELOCK;
```

Comment:
Ok, so we only clear this flag once all file locks on the inode have been released. What happens though if I open a new file descriptor on an inode that still has some file locks from before the blacklisting? Will I see errors on (e.g.) fsync on the new fd up until the point where all of the other filelocks are removed? That doesn't seem ideal if so.

ISTM that file descriptors opened after unblacklisting ought to work correctly, regardless of the state of open fd's that existed before blacklisting. Is that wrong?

Reply:
Looking closer, get_caps has this:

```
    if ((in->flags & I_ERROR_FILELOCK) && fh->has_any_filelocks())
      return -EIO;
```
Suppose the program has Fh `a` with a lock set on a file, and then we get a blacklisting/unblacklisting event. We set I_ERROR_FILELOCK, and then it opens Fh `b` and sets another (non-conflicting?) lock on the file. The program then closes Fh `a` but I_ERROR_FILELOCK is never cleared because Fh `b` was opened before `a` was closed.

Maybe we need a reconnect sequence counter or something so we can identify Fh's that were opened before the reconnect occurred?

Reply:
Fh b can't get new file lock before all stale file locks get released